### ISSUE TITLE
Allow control over automatic bond assignment

### ISSUE THREAD

**Issue by andrewtarzia** [lab-cosmo/chemiscope#361]
Hey all, This is working great so far. Except that my models are coarse-grained, so the bonds that come out from automatic assignment are not correct.

I have the topology information, so I can provide bonds (ideal solution), or I was think I could generate cylinder shapes for each bond. An alternative is that I could scale, or modify, the radii constrain for determining bonding information.

Any suggestions on which approach is best?

**Comment by ceriottm** [comment]
If you want to implement the machinery to provide bond information we'd be open to include it (though I'd suggest we first have a bit of discussion on how to do it so that it doesn't make other use cases more complicated or slow). We most definitely do not have the bandwidth to actively help though.

Otherwise, if you want to do this with existing infrastructure, manually defining the bonds as shapes is probably the easiest way, look at the implementation of `ase_merge_pi_frames`.

**Comment by andrewtarzia** [comment]
I would be very happy to contribute this feature in the future. I am not familiar with typescript, but it does seem like that would be quite a major overhaul of `dataset.ts` and `viewer.ts` to do this without breaking anything.

So, I will work with the `ase_merge_pi_frames` code for now, which seems to do exactly what I want with the shapes.

Thank you!

**Comment by ceriottm** [comment]
If I may make a suggestion - if you make a `topology_to_shapes` utility function that is fairly general and based on standard topo formats, this would be a very much welcome contribution. I'm leaving this open in case you go this way.

**Comment by Luthaf** [comment]
This would be quite nice to have. If someone wants to take a shot at it, here is what would need to happen for a full, clean integration:

- add bond information in the [`Structure`](https://github.com/lab-cosmo/chemiscope/blob/d162f876babae4a7f03ab907dded496d92a63ca2/src/dataset.ts#L88-L130) interface, likely as `Array[[number, number]]`, where each number is the index of one atom in the bonds
- use this bond information in the `StructureViewer`, using the current automatic bond assignment only if the data is not provided in the input.
- update the Python code to emit bond information when converting from user-provided structures. I don't think ASE stores the topology in the `Atoms` objects, so this would be a feature of other structure converters / manually added after conversion.

None of this should break other code, it should be fairly self-contained!

**Comment by andrewtarzia** [comment]
Thank you, both! 

I will start with the `topology_to_shape` approach because it seems simpler. 

Once I have that working, I can share that and we can see if @Luthaf suggestion would work nicely based on it.

Thank you again!

**Comment by ceriottm** [comment]
I think both options are useful to have so it's great to start from the easy one. Good chemiscoping.

**Comment by andrewtarzia** [comment]
Hey both,

I have managed to get bonding information written as shapes, but I am not seeing the shapes in the viewer, and I cannot figure out if there is something I am obviously doing wrong. Attached is a json file with one structure:

[cs_2p3_toff_2P3.json](https://github.com/user-attachments/files/16839821/cs_2p3_toff_2P3.json)

Because there are more bonds than atoms, I had to use the structure properties to add vectors. But then you need one entry per structure for a given shape, so I have N shapes for the number of bonds, which have a list of vectors and positions for each structure.

**Comment by ceriottm** [comment]
yes this definitely won't work, you're right. didn't think that the path integral helper has one bond per atom.
I can see two options (besides doing the "right thing" straight away and implement @Luthaf  solution). the "I really want a hack" way would be to create dummy atoms on which to pin the bonds, the other (with a complexity which is half-way) is to create a shape type that is a list of vectors so it can draw many cylinders per frame. Still, it's strange that the file you created does not show any bond at all, it should work.

**Comment by andrewtarzia** [comment]
I think, considering my time frame, it makes sense to just learn something new and implement the "right thing". I am not in a rush, and I like the idea of contributing to this project.

I will submit a PR, probably in draft form with something to discuss at some point.

I agree that it is strange that that file fails to show any shape.

**Comment by Luthaf** [comment]
So #363 should fix the (two) issues related to using cylinders to display bonds. 

> so I have N shapes for the number of bonds, which have a list of vectors and positions for each structure.

I can see two other workarounds here: (a) you can create a single "custom" shape instead of cylinders, containing all bonds in the system (by manually building a mesh corresponding to all the cylinders) or (b) you can add multiple shapes and then add `settings` in your JSON file to always select them all by default, which removes the need to select them for people viewing the dataset.

**Comment by andrewtarzia** [comment]
That is great, thank you! So based on that fix, the file I sent should work (with the addition of the named shapes in settings)?

**Comment by Luthaf** [comment]
Yes, this file works with the linked PR. If you checkout the branch locally, you can run `pip install .` and use it from Python, or just `npm start` to get the website interface & load your dataset there.

**Comment by andrewtarzia** [comment]
This works a charm - I am happy to contribute the topology to shapes code as a utility once I make it more general if you are all interested. I still intend on doing the more complete method @Luthaf described but it may take some time.

**Comment by andrewtarzia** [comment]
There is an issue in my code for the [example](https://chemiscope.org/docs/examples/9-showing_custom_bonds.html) of this.

I think the settings need to be added to the second jupyter visualisation.

**Comment by ceriottm** [comment]
@andrewtarzia I was having a look. I really like the possibility of doing custom bonds, and it's great that you have now something you can use in your project. I'd really like however (if you have the bandwitdh) not to close this yet, and try to move towards a "cleaner" solution - in particular one in which you don't have to define a separate shape per bond. 
I'm thinking of defining a `collection` shape type where you can define a base shape and then essentially add a dimension to the arrays that define the positions so you can make a single shape type that contains all bonds.

**Comment by andrewtarzia** [comment]
Hey both, I do agree. And now that I feel a bit more familiar with the underlying code, I would be willing to give it a go. I have some holidays coming up, so it won't be until after them. If it's ok to leave it open for now and I'll come back to it in a month or so?

**Comment by Luthaf** [comment]
Yes, this can wait!

**Comment by andrewtarzia** [comment]
Hey both, my bandwidth has reopened now (the job hunt is a pain) and I am keen to push forward on this. Reading through the comments above, it seems the plan is clear (as defined by @Luthaf) - to implement specific CustomBond attributes, that are parsed and handled by 3DMol.js. This should be able to take any list of "bonds" the user wants to provide.

If there are no other considerations or changes, I will get started and submit a Draft PR when ready for discussion? If there are significant changes to consider, I am happy to have a further chat!

**Comment by Luthaf** [comment]
Hey! The only other design consideration I can see would be bond order: do we want the end user to be able to specify that something is a single/double/… bond, and then try to render that? I'd say no for now, but I'm open for discussion!

**Comment by andrewtarzia** [comment]
I would say that bond order is worth it if it comes easily with 3DMol.js and the interface - I will set it as a goal and work backwards!

**Comment by andrewtarzia** [comment]
Sorry, I did not mean to close without comment/question - can we close this now regarding #416 ?

**Comment by ceriottm** [comment]
Yes I think there's a way to add it, and if someone wants another approach, then it's another issue and another PR ^^

**Comment by Luthaf** [comment]
So we now have both `convert_stk_bonds_as_shapes` and explicit bonds being passed through chemiscope's JSON file and rendered by 3Dmol.

Are you using both functions @andrewtarzia? We are pushing for a 1.0 release, and as part of that cleaning up the API of this package.

**Comment by andrewtarzia** [comment]
Hey @Luthaf , I do actually use both, but I presume that's just legacy code from me. I would say I like the utility to show non-molecular information (like in the stk example in the docs), but if you wanted to remove it or generalise it, then I can pull it out into my own code base.

**Comment by Luthaf** [comment]
You are talking about this example, right? https://chemiscope.org/docs/examples/9-showing-custom-bonds.html

I'm only seeing additional bonds being shown, am I missing something? 
It feels like these would be handled by this bit of code, right? 

https://github.com/lab-cosmo/chemiscope/blob/309adaae3235d1036a0e57b46f0d98f3958f30f0/python/chemiscope/structures/_stk.py#L56-L64

**Comment by andrewtarzia** [comment]
Yes, exactly. I believe the "orange bonds" (that show topological information of the cage) are from the `convert_stk_bonds_as_shapes`, but it is actually not the designed usage of that code because they are "fake" bonds. So, perhaps the function can be removed, and I can write a github gist showing how to do this for my own interest...

<img width="299" height="183" alt="Image" src="https://github.com/user-attachments/assets/f9e5cb2a-ffd5-49c7-8d40-2af6866490c9" />

**Comment by Luthaf** [comment]
Even there, because you are adding the bonds to the stk molecules just above it should be picked by the code I linked. The only missing thing would be the color specification.

If this works for you, I'd prefer to remove `convert_stk_bonds_as_shapes` then.

**Comment by andrewtarzia** [comment]
I am happy with this - good luck with the release!
